### ISSUE TITLE
[question/feature] Is it possible to upload recipe without creating?

### ISSUE THREAD

**Issue by Ahajha** [conan-io/conan#11924]
So to preface, this to me personally is a workaround for a missing feature in Gitlab's pipeline API, but I think it may have other uses.

For my org's usecase, we need to upload a recipe after we have verified it builds on all platforms. But we only upload the recipe, we never upload binaries. Due to Gitlab limitations that I won't go into detail on here(tl;dr we either have to up and download a lot of things, or recreate the package, both are extremely time consuming and wasteful), it would be convenient if we could just upload the recipe on its own without creating it.

I could see this being useful on its own, perhaps you're testing locally, you've already created a handful of test recipes and just want to upload the final product without the build artifacts, this would allow for that.

It would also be useful for us in dropping need for `conan copy`, as it's how we currently promote our packages. Instead, we could just upload the recipe with the new user/channel.

This isn't really a make or break feature, but it's one of our potential workarounds that would be useful to us. I'd love to hear any thoughts on this.

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @Ahajha 

> For my org's usecase, we need to upload a recipe after we have verified it builds on all platforms. But we only upload the recipe, we never upload binaries. 

You can upload only the recipe doing a ``conan upload <pkg-ref>`` without providing the ``--all`` and without providing the ``-p`` it will not upload any binaries. So for example, doing a ``conan export ...`` + ``conan upload <ref> -r=myremote`` will upload only the recipe.

> It would also be useful for us in dropping need for conan copy, as it's how we currently promote our packages. Instead, we could just upload the recipe with the new user/channel.

Yes, please note ``conan copy`` is deprecated and has been removed in 2.0. It is better not to use it. The recommended way to run promotions is to copy packages from one remote repo to another remote repo.

> Due to Gitlab limitations that I won't go into detail on here(tl;dr we either have to up and download a lot of things, or recreate the package, both are extremely time consuming and wasteful), it would be convenient if we could just upload the recipe on its own without creating it.

It is a bit surprising that re-building packages is as expensive as retrieving them, typically it is at least one order of magnitude slower. Specially when you have some packages that depend on other packages, re-building the transitive dependencies all the time just to build the consumers, is typically a big waste of time. If your server has some limitation, maybe you want to try with an ArtifactoryCE, it is completely free, and easy to run in your own server.

**Comment by Ahajha** [comment]
> doing a `conan export ...` + `conan upload <ref> -r=myremote` will upload only the recipe

This seems to be exactly what I'm looking for, I'm going to test that tomorrow and see how it goes, if it works I'll close this issue.

> Yes, please note conan copy is deprecated and has been removed in 2.0. It is better not to use it. The recommended way to run promotions is to copy packages from one remote repo to another remote repo.

I was aware of `conan copy` being removed in 2.0, hence mentioning it. Kills two birds with one stone (actually I counted 5 internal issues that this would solve, it's quite a problem for us haha). One of my coworkers looked into the package promotions on Gitlab, and he couldn't find a way.

> It is a bit surprising that re-building packages is as expensive as retrieving them, typically it is at least one order of magnitude slower.

The problem is that we're copying the cache, building into the copy, removing the other packages, and then uploading and downloading the remaining cache. It's not a good setup at all, and it's caused us some grief. It might be better if we just rebuilt on the upload (for most of our packages, at least), or now just doing `conan export` + `conan upload` I think would solve all our problems.

> If your server has some limitation, maybe you want to try with an ArtifactoryCE, it is completely free, and easy to run in your own server.

We're looking into this actually. Gitlab has a few missing features in its CI pipelines that would solve a lot of our headaches, and its Conan support isn't great. We can't even use version ranges, which would be a nice QOL improvement. If we can at least fix the Conan limitations, that would be a huge step in the right direction.
